### PR TITLE
Better handling of input prompts in first iteration

### PR DESCRIPTION
**Pull request by anwai98**
WIP!

### CODE REVIEW

**Comment by anwai98** [comment]
Okay, I can reproduce the error with `mask_prob=0.5` now.

```
Parameter at index 190 with name sam.prompt_encoder.mask_downscaling.6.weight has been marked as ready twice. This means that multiple autograd engine  hooks have fired for this particular parameter during this iteration.
```

DDP does not like this for some reason. I will investigate this now!

**Comment by anwai98** [comment]
Hi @constantinpape,

As discussed, this PR is ready as per our discussion. I checked it on multi-GPU setups, and works as expected!

Let me know how this looks. I will run a few training checks with it on LIVECell for now!

File: micro_sam/training/sam_trainer.py [hunk 1]
```diff
@@ -203,6 +203,41 @@ def _compute_iterative_loss(self, batched_inputs, y_one_hot, num_subiter, multim
...
+            if is_data_parallel:  # if training on multiple GPUs.
+                if torch.distributed.get_rank() == 0:  # device with rank 0.
+                    use_mask_inputs_tensor = torch.tensor(
+                        random.random() > self.mask_prob, dtype=torch.uint8, device=self.device,
+                    )
+                else:  # on other devices, we do not need this parameter at this stage.
+                    use_mask_inputs_tensor = torch.tensor(0, dtype=torch.uint8, device=self.device)
+
+                # Broadcast the value to all devices (ranks).
+                torch.distributed.broadcast(use_mask_inputs_tensor, src=0)
```
Comment (constantinpape) [inline micro_sam/training/sam_trainer.py:225]: Does this imply that we wait here for the variable to be available across iterations?
(I guess you checked / if the tests work then this means that it is correct.)

**Review by constantinpape — COMMENTED** [review]
Looks good to me! Let's wait to see if the training is working, and then I could merge this.

File: micro_sam/training/sam_trainer.py (hunk 1 above)
Comment (anwai98) [reply micro_sam/training/sam_trainer.py:225]: Ahha yes I forgot to address this. Yes, it indeed means that the "source" of this variable is from one particular device!

**Comment by anwai98** [comment]
hi @constantinpape,

As discussed yesterday, I have finalized this to keep our old single GPU heuristic for using mask inputs with a probability. GTG from my side!

**Review by constantinpape — COMMENTED** [review]
I think this is exactly what we discussed, but we need to delineate the two different cases of the mask usage a bit better in the code. It's maybe enough to add a few more comments, but right now I tend to separating it out a bit more. Let's discuss tomorrow.

File: micro_sam/training/sam_trainer.py [hunk 2]
```diff
@@ -195,6 +196,43 @@ def _get_best_masks(self, batched_outputs, batched_iou_predictions):
...
+
+        if self.mask_prob == 1:  # i.e. always use masks.
+            use_mask_inputs = True  # we would like to use mask inputs in all sub-iterations.
+            use_zero_mask = self.is_data_parallel  # we would like to use zeros as mask inputs for zeroth iteration.
+
+        elif self.mask_prob > 0:  # i.e. if we use mask inputs with a probability.
+            if self.is_data_parallel:  # if training on multiple GPUs.
+                if torch.distributed.get_rank() == 0:  # device with rank 0.
+                    use_mask_inputs_tensor = torch.tensor(
+                        random.random() > self.mask_prob, dtype=torch.uint8, device=self.device,
```
Comment (constantinpape) [inline micro_sam/training/sam_trainer.py:212]: The sign of the comparison is different from line 309. I think the one in 309 is correct, but please double check.

File: micro_sam/training/sam_trainer.py [hunk 3]
```diff
@@ -260,13 +301,21 @@ def _update_prompts(self, batched_inputs, y_one_hot, masks, logits_masks):
...
-                if use_mask_inputs:
-                    _inp["mask_inputs"] = logits
-                else:  # remove  previously existing mask inputs to avoid using them in next sub-iteration
-                    _inp.pop("mask_inputs", None)
+            if self.is_data_parallel:  # multi-GPU training
+                use_mask_inputs_this_iter = use_mask_inputs
+            else:  # single GPU training
+                if self.mask_prob > 0:
+                    # using mask inputs for iterative prompting while training, with a probability
+                    use_mask_inputs = (random.random() < self.mask_prob)
```
Comment (constantinpape) [inline micro_sam/training/sam_trainer.py:309]: Check previous comment.

**Review by constantinpape — COMMENTED** [review]
The overall logic should now be correct and is more clear, but there is an inconcistency in the random mask selection logic.

File: micro_sam/training/sam_trainer.py (hunk 2 above)
Comment (anwai98) [reply micro_sam/training/sam_trainer.py]: Yes, you are right. It's actually safe to stay consistent while comparing with the generator in two setups. Thanks for spotting this!

(I have fixed this in a new commit)
